### ISSUE TITLE
Clang should not choke on `#pragma GCC diagnostics ignored` it doesn't understand.

### ISSUE THREAD

**Issue by roccomoretti** [llvm/llvm-project#219294]
Minimal example (cut down from a much larger program where usage actually makes sense):

```c++
int main() {
#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
        return 0;
}
```

Compiling with Clang 23.1

```
clang++ test.cc -Werror
```

Results in

```
test.cc:2:32: error: unknown warning group '-Wmaybe-uninitialized', ignored [-Werror,-Wunknown-warning-option]
    2 | #pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
      |                                ^
1 error generated.
```

(GCC 16.2 can compile the program just fine, and obeys the directive in situations where it makes sense.)

Clang should not throw an error on this. While it's nice that Clang has support for `#pragma GCC diagnostic` functionality, those lines are intrinsically a directive to GCC. There are going to be warning options which GCC supports which Clang does not. Having Clang look at directives specifically intended for GCC and assuming it _always_ applies to Clang too is rather frustrating when you're trying to simultaneously support multiple compilers. (Which I assume is the goal of reading the `#pragma GCC` directives in the first place.) 

Note I'm only speaking of the `#pragma GCC diagnostic ignored` directive. I'm not attempting to say anything about command line options proper. If you're explicitly calling clang++ (or otherwise explicitly referring to Clang), it's reasonable to expect you're only ever going to pass options Clang understands. That's substantially different than situations where GCC-targeted directives (with no Clang references) are "helpfully" being picked up by Clang.

P.S. I would also file a feature request with GCC that they should really support `#pragma GCC_AND_BY_GCC_I_MEAN_GCC_ONLY_AND_NOT_CLANG_TOO`, but I don't think they'll go for it.

**Comment by shafik** [comment]
@AaronBallman @erichkeane I am not sure about the history on this feature here, I hoping some others do

**Comment by erichkeane** [comment]
We actually DONT emit an error on this, unless you specifically mention it.  We absolutely SHOULD warn here (else the pragma becomes horrifyingly unergonomic).  But as a now-retired implementer once said to me, "If you use -Werror and complain that it is preventing your build, you should be looking at yourself, not the compiler".

**Comment by AaronBallman** [comment]
`#pragma GCC diagnostic` and `#pragma clang diagnostic` are synonyms: https://clang.llvm.org/docs/UsersManual.html#controlling-diagnostics-via-pragmas and we don't want to try to track which warning groups GCC does or does not support, so it really boils down to "should the pragma tell you about unknown diagnostic groups?" to which the answer is "yes, same as the command line."

Btw, there is a workaround: https://godbolt.org/z/aoGeqx6j4 (keep in mind that with `_Pragma` you can also hide this stuff behind a macro to make it more ergonomic for highly portable code).

I think this issue should be closed as Won't Fix because it's behaving by design. Silencing the unknown warning group diagnostic in the GCC form would be possible, but it also means we would lose functionality that is useful like catching simple typos without having to compile with GCC to find them. (Also: because these are synonyms and have been for years, there are folks who compile only with Clang but they're using the GCC form of the pragma.)
